### PR TITLE
Remove local DMFiles and Checkpoints on failure

### LINKED ISSUE #7265: A DMFile can not be removed after apply snapshot meets decode error
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 3. What did you see instead (Required)
```
2023-04-08 15:06:56 | [2023/04/08 07:06:56.786 +00:00] [INFO] [ApplySnapshot.cpp:391] ["Decoding Region snapshot data meet error, sync schema and try to decode again [region 1070200342, applied: term 86 index 132] [error=DB::Exception: RegionBlockReader decode error]"] [thread_id=1085]Show context
-- | --
  |   | 2023-04-08 15:06:54 | [2023/04/08 07:06:54.211 +00:00] [DEBUG] [SSTFilesToDTFilesOutputStream.cpp:174] ["Create new DTFile for snapshot data, region=[region 1070200342, applied: term 86 index 132] file_idx=1 file=/data0/tiflash/data/ks_3490_t_1514/stable/.tmp.dmf_1238"] [source="table_id=1514"] [thread_id=1085]
  |   | 2023-04-08 15:06:54 | [2023/04/08 07:06:54.167 +00:00] [INFO] [SSTFilesToDTFilesOutputStream.cpp:212] ["Finished writing DTFile from snapshot data, region=[region 1070200342, applied: term 86 index 132] file_idx=0 file_rows=1007606 file_bytes=416634494 data_range=[39163459730,39311249079) file_bytes_on_disk=76945102 file=/data0/tiflash/data/ks_3490_t_1514/stable/dmf_1237"] [source="table_id=1514"] [thread_id=1085]
  |   | 2023-04-08 15:06:49 | [2023/04/08 07:06:49.018 +00:00] [DEBUG] [SSTFilesToDTFilesOutputStream.cpp:174] ["Create new DTFile for snapshot data, region=[region 1070200342, applied: term 86 index 132] file_idx=0 file=/data0/tiflash/data/ks_3490_t_1514/stable/.tmp.dmf_1237"] [source="table_id=1514"] [thread_id=1085]
  |   | 2023-04-08 15:06:48 | [2023/04/08 07:06:48.928 +00:00] [INFO] [SSTFilesToBlockInputStream.cpp:104] ["Finish Construct MultiSSTReader, write 1 lock 1 default 0 region 1070200342"] [source="table_id=1514"] [thread_id=1085]
...

2023-04-08 15:08:12 | [2023/04/08 07:08:11.128 +00:00] [INFO] [DeltaMergeStore_InternalBg.cpp:169] ["GC try remove useless DM file, but error happen, dmfile=/data0/tiflash/data/ks_3490_t_1514/stable/dmf_1237 err_msg=Check iter != pool.dt_file_path_map.end() failed: Cannot find DMFile when removing, file_id=1237"] [source="keyspace_id=3490 table_id=1514"] [thread_id=1102]Show context
-- | --
  |   | 2023-04-08 15:07:10 | [2023/04/08 07:07:10.434 +00:00] [INFO] [DeltaMergeStore_InternalBg.cpp:169] ["GC try remove useless DM file, but error happen, dmfile=/data0/tiflash/data/ks_3490_t_1514/stable/dmf_1237 err_msg=Check iter != pool.dt_file_path_map.end() failed: Cannot find DMFile when removing, file_id=1237"] [source="keyspace_id=3490 table_id=1514"] [thread_id=1099]
```

### 4. What is your TiFlash version? (Required)
master

### PR DESCRIPTION
**Pull request by JinheLin**
### What problem does this PR solve?

Issue Number: close #7265

### What is changed and how it works?
1. DMFile
  a. If exception happened when creating or uploading DMFiles, catch the exception and remove DMFiles.
  b. If in disagg-mode, removing all local DMFiles when restarting.
2. Checkpoint file
  a. Removing local checkpoint files when upload-checkpoint procedure finished whether successful or unsuccessful.
  b. Removing local checkpoint files when restarting.
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/DeltaMerge/SSTFilesToDTFilesOutputStream.cpp [hunk 1]
```diff
@@ -319,12 +319,11 @@ std::vector<ExternalDTFileInfo> SSTFilesToDTFilesOutputStream<ChildStream>::outp
 template <typename ChildStream>
 void SSTFilesToDTFilesOutputStream<ChildStream>::cancel()
 {
-    // Try a lightweight cleanup the file generated by this stream (marking them able to be GC-ed).
     for (auto & file : ingest_files)
     {
         try
         {
-            file->enableGC();
+            file->remove(context.getFileProvider());
```
Comment (lidezhu) [inline dbms/src/Storages/DeltaMerge/SSTFilesToDTFilesOutputStream.cpp:326]: The file's size may already be added into `StableDiskDelegator` in `SSTFilesToDTFilesOutputStream::writeSuffix`. And how to remove the size of this file?

**Review by lidezhu — APPROVED** [review]
LGTM

File: dbms/src/Storages/Page/V3/Universal/UniversalPageStorage.cpp [hunk 2]
```diff
@@ -499,24 +499,14 @@ PS::V3::CPDataDumpStats UniversalPageStorage::dumpIncrementalCheckpoint(const Un
...
-        tryLogCurrentException(log, "failed to persist checkpoint");
-        return {.has_new_data = false}; // TODO: maybe return has_new_data=true but upload_success=false?
-    }
+
+    auto checkpoint = PS::V3::LocalCheckpointFiles{
+        .data_files = data_file_paths,
+        .manifest_file = {manifest_file_path},
+    };
+    bool persist_done = options.persist_checkpoint(checkpoint);
+    RUNTIME_CHECK(persist_done);
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Page/V3/Universal/UniversalPageStorage.cpp:508]: why throw exception directly instead of returning `.has_new_data=false`?

**Review by JaySon-Huang — APPROVED** [review]
LGTM

**Pull request merged**